label_studio/io_storages/base_models.py

```diff
@@ -1,53 +1,235 @@
+    """
+    StorageInfo helps to understand storage status and progress
+    that happens in background jobs
+    """
+    class Status(models.TextChoices):
+        INITIALIZED = 'initialized', _('Initialized')
+        QUEUED = 'queued', _('Queued')
+        IN_PROGRESS = 'in_progress', _('In progress')
+        FAILED = 'failed', _('Failed')
+        COMPLETED = 'completed', _('Completed')
+
+    class Meta:
+        abstract = True
+
+    last_sync = models.DateTimeField(
+        _('last sync'),
+        null=True,
+        blank=True,
+        help_text='Last sync finished time'
+    )
+    last_sync_count = models.PositiveIntegerField(
+        _('last sync count'),
+        null=True,
+        blank=True,
+        help_text='Count of tasks synced last time'
+    )
+    last_sync_job = models.CharField(
+        _('last_sync_job'),
+        null=True,
+        blank=True,
+        max_length=256,
+        help_text='Last sync job ID'
+    )
+
+    status = models.CharField(
+        max_length=64,
+        choices=Status.choices,
+        default=Status.INITIALIZED,
+    )
+    traceback = models.TextField(
+        null=True,
+        blank=True,
+        help_text='Traceback report in case of errors'
+    )
+    meta = JSONField(
+        'meta',
+        null=True,
+        default=dict,
+        help_text='Meta and debug information about storage processes'
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # start datetime for statuses
+        self.time_in_progress = None
+        self.time_queued = None
+
+    def info_set_queued(self, job_id):
+        self.last_sync = None
+        self.last_sync_count = None
+        self.last_sync_job = job_id
+        self.status = self.Status.QUEUED
+
+        # reset and init meta
+        self.time_queued = datetime.now()
+        self.meta = {
+            'attempts': self.meta.get('attempts', 0) + 1,
+            'time_queued': str(self.time_queued)
+        }
+
+        self.save(update_fields=['last_sync_job', 'last_sync', 'last_sync_count', 'status', 'meta'])
+
+    def info_set_in_progress(self):
+        # only QUEUED => IN_PROGRESS transition is possible, because in QUEUED we reset states
+        if self.status != self.Status.QUEUED:
+            raise ValueError(f'Storage status ({self.status}) must be QUEUED to move it IN_PROGRESS')
+        self.status = self.Status.IN_PROGRESS
+
+        self.time_in_progress = datetime.now()
+        self.meta['time_in_progress'] = str(self.time_in_progress)
+        # at the very beginning it's the same as in progress time
+        self.meta['time_last_ping'] = str(self.time_in_progress)
+
+        self.save(update_fields=['status', 'meta'])
+
+    def info_set_completed(self, last_sync_count, **kwargs):
+        self.status = self.Status.COMPLETED
+        self.last_sync = timezone.now()
+        self.last_sync_count = last_sync_count
+
+        time_completed = datetime.now()
```

Comment:
Why we don't use `timezone.now()` for all cases?

Reply:
For one server you can do this, but for geo-distributed server configuration it's not the best approach, because you can have simultaneously running servers in different timezones, and if you take just local time, your time_completed calculation will be wrong.